### ISSUE TITLE
Elasticsearch queries with duplicated timeseries can block metrics endpoint

### ISSUE THREAD

**Issue by r4j4h** [braedon/prometheus-es-exporter#30]
Hi there,

I believe I have found a case where Prometheus occasionally reports `context deadline exceeded` while trying to scrape an instance of prometheus-es-exporter with many queries that take a range of time.

In particular I've noticed this one has some long-running queries (~4seconds) and if I repeatedly curl the /metrics endpoint I find it occasionally hangs before responding with the metrics.

I have not done enough in-depth analysis to prove anything but my hypothesis is that it is blocking while waiting for elasticsearch to respond so it can update its metrics, and while blocking Prometheus is timing out waiting for the response so it moves on until the next scrape cycle.

One fix is to increase the timeout that Prometheus will wait. Another is to accelerate Elasticsearch to respond faster to those queries, or to analyze the queries to see if they can be made more efficient and respond faster.

I am creating this issue because while those are obviously problems I should fix, I think that it is worth raising attention to this matter incase it affects others. The above mentioned strategies might work for them.

If my hypothesis is correct then at the cost of more complexity, threads or something similar to them might enable more responsive metrics responses during those long-running queries. I can see that these would be stale metrics though, and so perhaps it is better for it to fail in this case instead of fibbing. I also am not excited about asking to make an exporter more complex than it has to be.

Anyway, feel free to close this issue if users just need to correct it on their own, otherwise perhaps this can serve as a starting point for discussing a solution.

**Comment by r4j4h** [comment]
Just as a follow up, still encountering this but noticed that threads are already in use which for some reason I didn't before. 😊 

```
[2018-11-12 22:33:53,085] root.ERROR MainThread Error while running scheduled job.
Traceback (most recent call last):
  File "/usr/src/app/prometheus_es_exporter/__init__.py", line 210, in scheduled_run
    func()
  File "/usr/src/app/prometheus_es_exporter/__init__.py", line 121, in run_query
    update_gauges(metrics)
  File "/usr/src/app/prometheus_es_exporter/__init__.py", line 78, in update_gauges
    gauge = Gauge(metric_name, '', label_keys)
  File "/usr/local/lib/python3.7/site-packages/prometheus_client/core.py", line 623, in init
    registry.register(collector)
  File "/usr/local/lib/python3.7/site-packages/prometheus_client/core.py", line 59, in register
    duplicates))
ValueError: Duplicated timeseries in CollectorRegistry: {'nginx_request_durations_stats_min'}
```

I believe this duplicated timeseries error is the actual problem. It does seem to prevent responses to Prometheus scrapes though.

**Comment by r4j4h** [comment]
I have determined the cause for this was the use of `stats` in multiple places.

This is a partial from a configured query:

```

          "aggs": {
            "host": {
              "terms": {"field": "kubernetes.host.raw"},
              "aggs": {
                "namespace": {
                  "terms": {"field": "kubernetes.namespace_name.raw"},
                  "aggs": {
                    "pod": {
                      "terms": {"field": "kubernetes.pod_name.raw"},
                      "aggs": {
                        "duration": {  // <---- conflict
                          "percentiles": {"field": "duration"}
                        },
                        "stats": {   // <---- conflict
                          "stats": {"field": "duration"}
                        },
                        "statusCode": {
                          "terms": {"field": "status"},
                          "aggs": {
                            "method": {
                              "terms": {"field": "method.raw"},
                              "aggs": {
                                "method_stats": { // <---- a correction for conflict
                                  "stats": {"field": "duration"}
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            },
            "duration": {   // <---- conflict
              "percentiles": {"field": "duration"}
            },
            "stats": {   // <---- conflict
              "stats": {"field": "duration"}
            }
          }
```

After renaming the duplicate "stats" sections to have prefixes to prevent the collision, I am not seeing the error. And I am seeing the other stats populated in Prometheus. 👍

**Comment by r4j4h** [comment]
I am re-opening this issue because I believe these should not be conflicting after all, which is why I wrote it as I initially did.

An example patched rename was in the deepest aggregate on Method, `stats` -> `method_stats` which results in the time series:

`nginx_request_durations_host_namespace_pod_statusCode_method_method_stats_avg`

If this was just `stats` it should still be

`nginx_request_durations_host_namespace_pod_statusCode_method_stats_avg`

which is different from

`nginx_request_durations_stats_avg`

but perhaps the code is comparing the names before building out the full timeseries name? Just my working hypothesis. I am about to revert the config and test again.

**Comment by r4j4h** [comment]
After reverting the error does not immediately appear. Perhaps it is a long-running issue, or happened by having it run with one set of configuration and then some drift over time, but still keeping the collected metrics in its registry?

Kind of at a loss now.. will report back if I see the error re-occur.

**Comment by braedon** [comment]
Hi @r4j4h, based on what I believe is happening in https://github.com/braedon/prometheus-es-exporter/issues/36 to cause a similar symptom, my hypothesis is that the `min` value in the `stats` aggregation response is sometimes non-numeric (my best guess currently would be `None`). This can put the service in a bad state, resulting in a duplicate metric exception later on.

If so, this issue was probably fixed in [0.5.2](https://github.com/braedon/prometheus-es-exporter/releases/tag/0.5.2) by the fix for https://github.com/braedon/prometheus-es-exporter/issues/42. If you're not running it already, could you upgrade to this version, and confirm if you're still seeing the issue?

If you are stilling having the issue, could you let me know the mapping type of the `duration` field? It'd also be good to know if you ever get duplicate metric exceptions for any metrics other than `nginx_request_durations_stats_min`.

Thanks!
